README.md
# iLO Cookbook

[![Cookbook Version](https://img.shields.io/cookbook/v/ilo.svg)](https://supermarket.chef.io/cookbooks/ilo)
[![Travis Build Status](https://travis-ci.org/HewlettPackard/ilo-chef.svg?branch=master)](https://travis-ci.org/HewlettPackard/ilo-chef)
[![Chef Build Status](https://jenkins-01.eastus.cloudapp.azure.com/job/ilo-cookbook/badge/icon)](https://jenkins-01.eastus.cloudapp.azure.com/job/ilo-cookbook/)

Enables configuration of HPE iLOs via their APIs.

### Requirements

- Chef 12.7+
- iLO 4

### Cookbook Dependencies

- none

### How to use the iLO Cookbook:

This cookbook is not intended to include any recipes.
Use it by specifying a dependency on this cookbook in your own cookbook.

```ruby
# my_cookbook/metadata.rb
...
depends 'ilo'
```

Now you can use the resources this cookbook provides. See below for some examples.


## iLO Authentication

Each of the resources below requires you to pass in the info necessary to connect with the iLO API.
The basic structure accepted by the `ilos` property is an array of hashes (or `ILO_SDK::Client` objects):

```ruby
ilos = [
  {
    host: 'ilo1.example.com',  # Required. IP or hostname
    user: 'Administrator',     # Optional. Defaults to 'Administrator'
    password: 'secret123',     # Required
    ssl_enabled: false         # Optional
  },
  {
    host: '10.0.0.3',
    user: 'User2',
    password: 'secret456'
  }
]
```

This array can be built using a variety of different sources, including [encrypted] databags, attributes, or read from json or yaml files.
For example:

```ruby
# Set directly in recipe:
ilo_list1 = []
ilo_list1.push { host: 'ilo1.example.com', user: 'Administrator', password: 'secret123' }

# Read from data_bag:
ilo_list2 = data_bag_item('ilo_secrets', 'data_center_1')

# Load from yaml file:
ilo_list3 = YAML.load_file('/root/ilo_secrets.yml')
```


## iLO Resources

The following resources are available for usage in your recipes.
We give examples to show how to use each resource, but there are much more detailed examples in the [examples](examples) directory.
You may also find the [recipes in the cookbook used for testing](spec/fixtures/cookbooks/ilo_test/recipes) helpful as examples.

### ilo_bios

 - ####Set BIOS configuration:

  ```ruby
  ilo_bios 'set BIOS configuration' do
    ilos [ilo1, ilo2]
    settings(
      UefiShellStartup: 'Enabled',
      UefiShellStartupLocation: 'Auto',
      UefiShellStartupUrl: 'http://www.uefi.nsh',
      Dhcpv4: 'Enabled',
      Ipv4Address: '10.1.1.0',
      Ipv4Gateway: '10.1.1.11',
      Ipv4PrimaryDNS: '10.1.1.1',
      Ipv4SecondaryDNS: '10.1.1.2',
      Ipv4SubnetMask: '255.255.255.0',
      UrlBootFile: 'http://www.urlbootfile.iso',
      ServiceName: 'iLO Admin',
      ServiceEmail: 'admin@domain.com'
      # NOTE: This is just an example; you can set as many or as few settings as you want.
      # (And there's a whole lot more you can set. See the API docs for the complete list.)
    )
    action :set # Not necessary, as this is the default
  end
  ```

 - ####Revert to default BIOS base configuration:

  ```ruby
  ilo_bios 'revert BIOS' do
    ilos [ilo1, ilo2]
    action :revert
  end
  ```


### ilo_boot_settings

 - ####Set boot configuration:

  ```ruby
  ilo_boot_settings 'set boot configuration' do
    ilos [ilo1, ilo2]
    boot_order [       # Optional
      "FD.Virtual.1.1",
      "Generic.USB.1.1",
      "HD.Emb.1.1",
      "HD.Emb.1.2",
      "NIC.LOM.1.1.IPv4",
      "NIC.LOM.1.1.IPv6"
    ]
    boot_target 'None' # Optional
    action :set        # Not necessary, as this is the default
  end
  ```

 - ####Revert to default boot configuration:

  ```ruby
  ilo_boot_settings 'revert boot configuration' do
    ilos [ilo1, ilo2]
    action :revert
  end
  ```

 - ####Dump the boot configuration to a file:

  ```ruby
  ilo_boot_settings 'dump boot configuration' do
    ilos [ilo1, ilo2]
    dump_file '/full/path/to/boot_settings.yml'
    owner 'JohnDoe'        # Optional: Owner of the file. Defaults to the current user
    group 'Administrators' # Optional: Group ownership for the file. Defaults to the current user
    action :dump
  end
  ```


### ilo_chassis

 - ####Dump power metrics and thermal metrics information to a file:

  ```ruby
  ilo_chassis 'dump power metrics and thermal metrics' do
    ilos [ilo1, ilo2]
    power_metrics_file '/full/path/to/power_metrics.yml'
    thermal_metrics_file '/full/path/to/thermal_metrics.yml'
    owner 'JohnDoe'        # Optional: Owner of the file. Defaults to the current user
    group 'Administrators' # Optional: Owner of the file. Defaults to the current user
    action :dump           # Not necessary, as this is the default
  end
  ```


### ilo_computer_details

 - ####Dump computer details to a file and data bag:

  ```ruby
  ilo_computer_details 'dump computer details' do
    ilos [ilo1, ilo2]
    dump_file '/full/path/to/computer_details.yml'
    data_bag 'computer_details_bag'
    owner 'JohnDoe'        # Optional: Owner of the file. Defaults to the current user
    group 'Administrators' # Optional: Owner of the file. Defaults to the current user
    action :dump           # Not necessary, as this is the default
  end
  ```


### ilo_computer_system

 - ####Set computer system information:

  ```ruby
  ilo_computer_system 'set computer system info' do
    ilos [ilo1, ilo2]
    asset_tag 'HPE001' # Optional
    led_state 'Lit'    # Optional: 'Lit' or 'Off'
    action :set        # Not necessary, as this is the default
  end
  ```


### ilo_date_time

 - ####Set the time zone and NTP settings:

  ```ruby
  ilo_date_time 'set time zone' do
    ilos [ilo1, ilo2]
    use_dhcpv4 false           # Optional. Set to true to use the DHCPv4-supplied time settings
    time_zone 'Africa/Abidjan' # Optional. To use this, you must set :use_dhcpv4 to false
    ntp_servers [              # Optional. To use this, you must set :use_dhcpv4 to false
      "10.168.0.2",
      "10.168.0.3"
    ]
    action :set                # Not necessary, as this is the default
  end
  ```


### ilo_firmware_update

 - ####Upgrade firmware:

  ```ruby
  ilo_firmware_update 'upgrade firmware' do
    ilos [ilo1, ilo2]
    fw_version '2.51'
    fw_uri 'www.firmware.domain.com/2.51'
    action :upgrade # Not necessary, as this is the default
  end
  ```


### ilo_https_cert

Note that this resource requires an `ilo` property (Hash or ILO_SDK::Client) instead of an `ilos` property (Array).
You'll need separate ilo_https_cert resources for each iLO you'd like to perform a task on.

 - ####Generate Certificate Signing Request (CSR):

  ```ruby
  ilo_https_cert 'generate CSR' do
    ilo ilo1
    country 'US'
    state 'Texas'
    city 'Houston'
    orgName 'Example Company'
    orgUnit 'Example'
    commonName 'example.com'
    action :generate_csr
  end
  ```

 - ####Dump CSR to a file:

  ```ruby
  ilo_https_cert 'dump CSR to file' do
    ilo ilo1
    file_path '/full/path/to/CSR.cert'
    action :dump_csr
  end
  ```

 - ####Import certificate:

  ```ruby
  ilo_https_cert 'import certificate' do
    ilo ilo1
    certificate '-----BEGIN CERTIFICATE-----...'
    action :import # Not necessary, as this is the default
  end
  ```

 - ####Import certificate from file:

  ```ruby
  ilo_https_cert 'import certificate from file' do
    ilo ilo1
    file_path '/full/path/to/certificate_file.cert'
    action :import # Not necessary, as this is the default
  end
  ```


### ilo_log_entry

 - ####Dump log entries to a file:

  ```ruby
  ilo_log_entry 'dump log entries' do
    ilos [ilo1, ilo2]
    dump_file '/full/path/to/IEL_logs.txt'
    log_type 'IEL'         # 'IEL' or 'IML
    owner 'JohnDoe'        # Optional: Owner of the dump file. Defaults to the current user
    group 'Administrators' # Optional: Owner of the dump file. Defaults to the current user
    duration 30            # Optional: Number of hours ago to begin collection at (until now)
    severity_level 'OK'    # Optional: Exclude this property to get all severities
    action :dump           # Not necessary, as this is the default
  end
  ```

 - ####Clear log entries:

  ```ruby
  ilo_log_entry 'clear log entries' do
    ilos [ilo1, ilo2]
    log_type 'IEL'
    action :clear
  end
  ```


### ilo_manager_network_protocol

 - ####Set ilo session timeout:

  ```ruby
  ilo_manager_network_protocol 'set timeout' do
    ilos [ilo1, ilo2]
    timeout 60  # Minutes
    action :set # Not necessary, as this is the default
  end
  ```


### ilo_power

 - ####Power on the system:

  ```ruby
  ilo_power 'power on' do
    ilos [ilo1, ilo2]
    action :poweron # Not necessary, as this is the default
  end
  ```

 - ####Power off the system:

  ```ruby
  ilo_power 'power off' do
    ilos [ilo1, ilo2]
    action :poweroff
  end
  ```

 - ####Reset/Restart the system:

  ```ruby
  ilo_power 'reset system' do
    ilos [ilo1, ilo2]
    action :resetsys
  end
  ```

 - ####Reset/Restart the iLO:

  ```ruby
  ilo_power 'reset ilo' do
    ilos [ilo1, ilo2]
    action :resetilo
  end
  ```


### ilo_secure_boot

 - ####Set whether or not to enable UEFI secure boot:

  ```ruby
  ilo_secure_boot 'enable secure boot' do
    ilos [ilo1, ilo2]
    enable true # Optional: Defaults to false
    action :set # Not necessary, as this is the default
  end
  ```


### ilo_service_root

 - ####Dump schema and registry information to a file:

  ```ruby
  ilo_service_root 'dump schema and registry' do
    ilos [ilo1, ilo2]
    schema_prefix 'Account'
    schema_file '/full/path/to/schema.txt'
    registry_prefix 'Base'
    registry_file '/full/path/to/registry.txt'
    owner 'JohnDoe'        # Optional: Owner of the dump file(s). Defaults to the current user
    group 'Administrators' # Optional: Owner of the dump file(s). Defaults to the current user
    action :dump           # Not necessary, as this is the default
  end
  
  # Note: You must set the :schema_file or :registry_file property, but are not required to set both
  ```


### ilo_snmp_service

 - ####Configure SNMP service:

  ```ruby
  ilo_snmp_service 'set snmp mode and turn alerts on' do
    ilos [ilo1, ilo2]
    snmp_mode 'Agentless' # Optional: Defaults to 'Agentless'
    snmp_alerts true      # Optional: Defaults to  false
    action :configure     # Not necessary, as this is the default
  end
  ```


### ilo_user

 - ####Create or modify user:

  ```ruby
  ilo_user 'create user' do
    ilos [ilo1, ilo2]
    username 'test'                   # Defaults to the resource's name attribute
    password 'password123'            # Optional for updates, but needed to create
    login_priv true                   # Optional
    remote_console_priv false         # Optional
    user_config_priv true             # Optional
    virtual_media_priv false          # Optional
    virtual_power_and_reset_priv true # Optional
    ilo_config_priv false             # Optional
    action :create                    # Not necessary, as this is the default
  end
  ```

 - ####Delete user:

  ```ruby
  ilo_user 'delete user' do
    ilos [ilo1, ilo2]
    username 'test'
    action :delete
  end
  ```


### ilo_virtual_media

 - ####Insert virtual media:

  ```ruby
  ilo_virtual_media 'insert virtual media' do
    ilos [ilo1, ilo2]
    iso_uri 'http://10.254.224.38:5000/ubuntu-15.04-desktop-amd64.iso'
    action :insert # Not necessary, as this is the default
  end
  ```

 - ####Eject virtual media:

  ```ruby
  ilo_virtual_media 'eject virtual media' do
    ilos [ilo1, ilo2]
    action :eject
  end
  ```


## Examples

See the [examples](examples) directory for examples with more detailed descriptions of using these resources.
It may also be helpful to take a look at [recipes in the cookbook used for testing](spec/fixtures/cookbooks/ilo_test/recipes).


## Contributing & Feature Requests

**Contributing:** Please see [CONTRIBUTING.md](CONTRIBUTING.md) for more info.

**Feature Requests:** If you have a need that is not met by the current implementation, please let us know (via a new issue).
This feedback is crucial for us to deliver a useful product. Do not assume we have already thought of everything, because we assure you that is not the case.

### License

This project is licensed under the Apache 2.0 license. Please see [LICENSE](LICENSE) for more info.

### Testing

Please see [CONTRIBUTING.md](CONTRIBUTING.md) for more info.

## Authors

 - Anirudh Gupta - [@Anirudh-Gupta](https://github.com/Anirudh-Gupta)
 - Bik Bajwa - [@bikbajwa](https://github.com/bikbajwa)
 - Jared Smartt - [@jsmartt](https://github.com/jsmartt)
 - Vivek Bhatia - [@vivekbhatia14] (https://github.com/vivekbhatia14)


CHANGELOG.md
### 1.4.0
 - Fixed bugs in ilo_date_time resource and renamed `use_ntp` property to `use_dhcpv4` (breaking change).
 - Remove dependency on compat_resource and bumped minimum Chef client version to 12.7
 - Added examples

### 1.3.1
 - Fixed bug in ilo_https_cert resource: Accept `ILO_SDK::Client` object for ilo property.
 - Fixed bug in ilo_firmware_update resource: Accept String for fw_version property.
 - Fix other various property restrictions,
 - Fixed bug in ilo_https_cert :import when importing from file.

### 1.3.0
 - Refactored ilo_bios resource to support all settings. (Breaking Change)

### 1.2.0
 - Moved resources into libraries
 - Added ilo_https_cert resource

#### 1.0.1
 - Updated metadata.rb file

## 1.0.0
Initial release


recipes/default.rb
#
# Cookbook Name:: iLORestCookbook
# Recipe:: default
#
# Copyright (c) 2015 The Authors, All Rights Reserved.
# Chef::Resource::Execute.send(:include, RestAPI::Helper)
# To Add user on all the
#

# iLO_user 'user reset password' do
#   username 'pappu'
#   password 'password12'
#   ilo_names ["ILO-02"]
#   action :changePassword
# end

# iLO_user 'user delete' do
#   username 'test'
#   ilo_names :all
#   action :deleteUser
# end


# iLO_user 'user create' do
#   username 'ram'
#   password 'password123'
#   ilo_names ["ILO-02"]
# end


# iLO_powermgmt 'power off' do
#   ilo_names :all
#   action :poweroff
# end
#
# iLO_powermgmt 'power on' do
#   ilo_names :all
#   action :poweron
# end


#  iLO_powermgmt 'resetsys' do
#    ilo_names :all
#    action :resetsys
#  end

# iLO_fw_up 'fw_up' do
#   ilo_names ["ILO-03"]
#   fw_version 2.3
#   fw_uri "http://10.254.224.38:8000/ilo4_240.bin"
#   action :fw_up
# end

# iLO_license 'apply license' do
#   ilo_names ["ILO-01"]
#   license_key "333TJ-XN732-5CRYY-RVXYH-KJDJR"
#   action :apply
# end

# iLO_logs 'clear logs' do
#   ilo_names ["ILO-02"]
#   log_type 'iel'
#   action :clear
# end

# iLO_logs 'dump iel logs' do
#   ilo_names ["ILO-02"]
#   log_type 'iml'
#   duration_in_hours 24
#   dump_file "iml_logs"
#   action :dump
# end
#
# iLO_uefi 'secure boot' do
#   ilo_names ["ILO-02"]
#   enable true
#   action :secure_boot
# end

# iLO_powermgmt 'resetsys' do
#   ilo_names ["ILO-02"]
#   action :resetsys
# end

# iLO_boot_order 'revert boot order' do
#   ilo_names ["ILO-02"]
#   action :revert
# end

# iLO_time_zone 'use NTP servers' do
#   ilo_names ["ILO-02"]
#   value false
#   action :use_ntp
# end
#

# iLO_time_zone 'set time zone' do
#   ilo_names ["ILO-02"]
#   time_zone "Asia/Kolkata"
#   action :set_time_zone
# end
#
# iLO_powermgmt 'resetilo' do
#   ilo_names ["ILO-02"]
#   action :resetilo
# end

# iLO_indicator_led 'set led state' do
#   ilos [ilo1]
#   led_state "Off"
# end

# iLO_computer_details 'dump computer details' do
#   ilo_names :all
#   dump_file 'comp_details1'
#   action :dump
# end

# iLO_virtual_media 'mount iso' do
#   ilo_names ['ILO-02']
#   iso_uri 'http://10.254.224.38:5000/ubuntu-15.04-desktop-amd64.iso'
#   boot_on_next_server_reset false
#   action :mount
# end


# iLO_asset_tag 'set asset tag' do
#   ilo_name 'ILO-02'
#   asset_tag 'HP001'
#   action :set
# end

# iLO_snmp 'configure snmp' do
#   ilo_names ['ILO-01']
#   snmp_mode 'Agentless'
#   snmp_alerts false
#   action :configure
# end

# iLO_registry 'get registry' do
#   ilo_name 'ILO-01'
#   registry_prefix 'iLO'
#   registry_file 'registry'
#   action :get_registry
# end


# iLO_schema 'get schema' do
#   ilo_name 'ILO-01'
#   schema_prefix 'ComputerSystem'
#   schema_file 'schema'
#   action :get_schema
# end

# iLO_boot_order 'get  boot order' do
#   ilo_names ['ILO-02']
#   boot_order_file 'boot_order_2'
#   action :get
# end

# iLO_boot_order 'change to new boot order' do   ilo_name 'ILO-02'
#   new_boot_order ["File.URL.1.1","HD.Emb.1.2", "Generic.USB.1.1","HD.Emb.1.1","NIC.LOM.1.1.IPv4","NIC.LOM.1.1.IPv6"]
#   action :change
# end
#
# iLO_powermgmt 'resetsys' do
#   ilo_names ['ILO-01']
#   action :resetsys
# end

# iLO_boot_order 'change temporary boot order' do
#   ilo_names ['ILO-01']
#   boot_target 'Hdd'
#   action :temporary_change
# end

# iLO_bios 'set uefi shell startup' do
#   ilo_names ['ILO-02']
#   value 'Enabled'
#   location 'Auto'
#   url 'http://hp.com/test.nsh'
#   action :uefi_shell_startup
# end
#
# iLO_powermgmt 'resetsys' do
#   ilo_names ['ILO-02']
#   action :resetsys
# end

# iLO_bios 'disable dhcp' do
#   ilo_names ['ILO-02']
#   value 'Disabled'
#   ipv4_address '192.168.0.1'
#   ipv4_gateway '192.168.0.2'
#   ipv4_primary_dns '192.168.0.3'
#   ipv4_secondary_dns '192.168.0.4'
#   ipv4_subnet_mask '192.168.0.5'
#   action :set_dhcp
# end

# iLO_powermgmt 'resetsys' do
#   ilo_names ['ILO-02']
#   action :resetsys
# end

# iLO_bios 'set bios url boot file' do
#   ilo_names ['ILO-02']
#   url_boot_file "http://hp.com/test.efi"
#   action :set_url_boot_file
# end

# iLO_bios 'set bios service' do
#   ilo_names ['ILO-02']
#   service_name "HPE"
#   service_email "anirudhg@hpe.com"
#   action :set_bios_service
# end

# iLO_bios 'revert bios settings' do
#   ilo_names ['ILO-02']
#   action :revert
# end



# iLO_powermgmt 'resetsys' do
#   ilo_names ['ILO-02']
#   action :resetsys
# end

# iLO_time_zone 'set ntp servers' do
#   ilo_names ["ILO-02"]
#   ntp_servers ['192.168.0.1', '192.168.0.2']
#   action :set_ntp_servers
# end
